File: benchmarks/vm/18_union_struct_payload_loop.rb
Point = Struct.new(:x, :y)
Box = Struct.new(:p, :weight)
Boxed = Struct.new(:box)
Raw = Struct.new(:n)
Empty = Object.new

def make_item(i)
  case i % 3
  when 0
    Boxed.new(Box.new(Point.new(i, i * 2), i % 7))
  when 1
    Raw.new(i * 5)
  else
    Empty
  end
end

def score_item(item)
  case item
  when Boxed
    item.box.p.x + item.box.p.y + item.box.weight
  when Raw
    item.n - 3
  else
    1
  end
end

t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
total = 0
100_000.times do |i|
  total += score_item(make_item(i))
end
total_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000
puts "total = #{total}"
puts format("BENCH_RESULT: %.3f ms", total_ms)

File: benchmarks/vm/18_union_struct_payload_loop.lua
local function make_item(i)
    local r = i % 3
    if r == 0 then
        return { tag = "Boxed", box = { p = { x = i, y = i * 2 }, weight = i % 7 } }
    elseif r == 1 then
        return { tag = "Raw", n = i * 5 }
    end
    return { tag = "Empty" }
end

local function score_item(item)
    if item.tag == "Boxed" then
        return item.box.p.x + item.box.p.y + item.box.weight
    elseif item.tag == "Raw" then
        return item.n - 3
    end
    return 1
end

local t0 = os.clock()
local total = 0
for i = 0, 99999 do
    total = total + score_item(make_item(i))
end
local total_ms = (os.clock() - t0) * 1000
print("total = " .. tostring(total))
print(string.format("BENCH_RESULT: %.3f ms", total_ms))
